### ISSUE TITLE
TabCompletion for plugins

### ISSUE THREAD

**Issue by madmaxoft** [cuberite/cuberite#701]
The plugins should be allowed to ~~interfere~~ interface with the tabcompletion so that they can provide their own results. Also the internal tabcompletion methods should be exported so that the plugins may use them.

**Comment by Seadragon91** [comment]
I think because of InfoReg.lua, we don't need that anymore, or?

**Comment by NiLSPACE** [comment]
No, we still want plugins to be able to change the tab completion. What if I wanted instead of a command or a player something else related to that plugin. For example `/gal visit P<tab>` would turn into `/gal visit Plains`

**Comment by Seadragon91** [comment]
Ah okay. Thanks you. Still something not correct with TAB completion. If I type `/protection a` and TAB, then the chat content is `/protection /protection add`

**Comment by williamhatcher** [comment]
Bump. As a plugin developer, this would be awesome

**Comment by madmaxoft** [comment]
@williamhatcher Can you propose what the API should look like? Designing a good API is one of the harder parts of development :)

**Comment by williamhatcher** [comment]
# Tab Completion
## `Info.lua`
```lua
...
["/somecommand"] = {
  Handler = HandleSomeCommand,
  TabCompletion = SomeCommandTabCompletion, -- function(a_Split, a_Player)
},
...
```
## TabCompletion Function
It should return a table of tables (mostly)
Ex: `{"/somecommand", {"internal option", "something else"}, {}, "players", "worlds"}`
The table should have the command name first, then tables containing possible results, empty tables to indicate no more possible results for that parameter, the string `players` for matching players, and the string `world` for matching worlds.
I think it will be more performant  if Cuberite itself populated the players and worlds, rather than the plugin having to load each one into a list.

## Cuberite's Handleing
- `TabCompletion` should be called each time a player presses tab (in chat)
- Cuberite should match each parameter against its corresponding table and replace that parameter with one from the table that matches it, and continues to cycle through partial matches
- If the first entry in the table is different from the command the player has typed, either:
  - Ignore that result, *or*
  - Change the command the player is trying to run.

**Comment by bearbin** [comment]
I think this is a lot of needless complexity - what about a the TabCompletion callback simply being given the split and returning a table of possible completions?

Cuberite could provide some utility functions (`TabCompleteWorlds(a_NameStart)`, `TabCompletePlayers(a_NameStart)`, etc.) that plugins would use if they had to complete those objects.

**Comment by williamhatcher** [comment]
Player should be given so plugins can look at permissions

**Comment by madmaxoft** [comment]
The `TabCompletion` callback should receive both the split and the source (player / `nil` for console (?)), and it should return only the list of possibilities that is relayed directly to the client; but that's only an implementation detail which will be mostly handled by the $/Server/Plugins/InfoReg.lua helper script (just like it handles the command registration itself).

The real question is about the set of API functions that Cuberite needs to provide (like `cPluginManager:BindCommand` is used for the actual command binding) and how the tab completion handlers will be registered, whether there'll be a separate hook so that a plugin can access tab completion for other things than commands, etc. That's what I meant by the API. But I have to say that the Info.lua integration sounds like a great idea, I haven't thought of that.

**Comment by williamhatcher** [comment]
I thought the console doesn't have any tab completion?

`BindCommand` could take a fifth parameter `TabCompletion` as function, which gives backwards compatibility.

>whether there'll be a separate hook so that a plugin can access tab completion for other things than commands, etc.

I'm not following. Can you please give me an example?

**Comment by madmaxoft** [comment]
There are things that plugins might want to tab-complete even when not specifying a command (in regular chat, for example). For these, you'd need a way to intercept the default tab completion from a plugin without having a command registered.

For example a "hide" plugin that a player can activate and it removes their nickname from the tab completion of other plugins (possibly from regular chat, too).

**Comment by madmaxoft** [comment]
Console doesn't have tab completion for now, but it may get one in the future.
